### ISSUE TITLE
[PROPOSAL] Protobuf Automation 

### ISSUE THREAD

**Issue by amberzsy** [opensearch-project/opensearch-api-specification#653]
### What are you proposing?
There are bunch gaps for having E2E protobuf and API spec release: 
1. Open api spec not in sync with core API and lack of automation and enforcement 
2. No automation and tooling for protobuf generation from api spec. 
3. Lack of validation on protobuf and api spec 

#### Development flow

- There will be a new repo called opensearch-api-protobufs, it can be nested as a submodule into opensearch core repo or separate repo similar to opensearch-api-specification 

- As part of grpc new api development,  developer manually update the protobuf repo, and update their local submodule configs to point to their protobuf changes. If it’s for adding corresponding grpc endpoint per http api, use https://github.com/dblock/opensearch-api to produce API spec at runtime and trigger spec<>protobuf conversion to generate protobuf definition. 

- During PR on opensearch core repo, there would be multiple workflow/job triggered on merge action, 

Workflow 1: Trigger spec generation job with versioning annotations, it would generate a PR in opensearch-api-spec repo. 

- Tooling: https://github.com/dblock/opensearch-api

Workflow 2: trigger ‘merge to protobuf’ which generate protobuf from the spec. 
Tooling: spec <> proto generator (WIP)

#### Linter support 

- Check the protobuf which is manually written is matched to what it will be generated by the spec. It would help with

1. Developer needs to manually write proto for grpc endpoint, linter would make sure it’s aligned with spec<>proto conversion rules.
2. Automatically generate Spec -> protos for the PR
3. Block the PR if manual proto definition does not align with Spec. 
4. Corner cases: api only support in grpc or http endpoint.

**Comment by dblock** [comment]
Thanks for opening this. 

I think we should resolve what the source of truth for OpenSearch would be with this proposal. My understanding was (still is :)) that the spec in this repo is the OpenSearch API truth. While today it's reverse engineering the existing OpenSearch API because of the legacy, we see a future where the OpenSearch server-side API is generated from spec as well. If this is true, then protobufs are just another output from the source of truth, aka a build of API spec in this repo that outputs .proto files.

WDYT?

**Comment by amberzsy** [comment]
> Thanks for opening this.
> 
> I think we should resolve what the source of truth for OpenSearch would be with this proposal. My understanding was (still is :)) that the spec in this repo is the OpenSearch API truth. While today it's reverse engineering the existing OpenSearch API because of the legacy, we see a future where the OpenSearch server-side API is generated from spec as well. If this is true, then protobufs are just another output from the source of truth, aka a build of API spec in this repo that outputs .proto files.
> 
> WDYT?
> 
> I opened #655 to express this.

open-api-spec in general is for REST/JSON, that's why propose have separate repo for proto. but not strong opinionated on this. i think we can start with keeping in same repo to start with. But we still need to close the gap that, for protobuf, server side development has to define the protobuf first, which is unlike json, which we can do reverse engineering. That's why introduce the **development flow** above.  

reg if spec the source of truth. imo, partially yes. for API schema, most likely, but for rpc etc, it would be different. another example would be, grpc response is very different from http. so most of response we cannot adopt or directly convert from spec. Or, .... we can make spec can support both protobuf and json with certain limitation etc.  

Also, actually, let me break into separate proposals.
1. proposal on how we can keep core <> spec in sync. (how and plan to fix the existing and stop gap). 
2. proposal on the protobuf conversion. (include tooling and the rules we build).
3. proposal on protobuf release/development cycle. (maybe can merge into 1. )

**Comment by dblock** [comment]
I suppose I am still stuck in the idea that the spec is the only API, but it sounds like you want the Protobuf spec to be potentially very different? Let's think from a developer POV (related to 1), if I am adding a new API, say "_refresh", how do you see one do it in a world where it's both a REST API and a protobuf API?

**Comment by guptashubham** [comment]
i agree with DB that spec is the only source of truth. However, spec needs to evolve as it will require versioning support (with the binary) among other things. I'd also expect spec to be the source of truth for API documentation and we can build it over time.
